libs/langchain-community/src/llms/googlegenerativeai.ts

```diff
@@ -0,0 +1,380 @@
+          "GoogleGenerativeAI constructor"
+      );
+    }
+
+    this.safetySettings = fields?.safetySettings ?? this.safetySettings;
+    if (this.safetySettings && this.safetySettings.length > 0) {
+      if (isGenerateContentModel) {
+        assertSafetySettings(
+          this.safetySettings as SafetySetting[],
+          HarmCategory,
+          HarmBlockThreshold
+        );
+      } else {
+        assertSafetySettings(
+          this
+            .safetySettings as protos.google.ai.generativelanguage.v1beta2.ISafetySetting[],
+          protos.google.ai.generativelanguage.v1beta2.HarmCategory,
+          protos.google.ai.generativelanguage.v1beta2.SafetySetting
+            .HarmBlockThreshold
+        );
+      }
+    }
+
+    this.client = isGenerateContentModel
+      ? new GenerativeAI(this.apiKey).getGenerativeModel({
+          model: this.modelName,
+          safetySettings: this.safetySettings as SafetySetting[],
+          generationConfig: {
+            candidateCount: 1,
+            stopSequences: this.stopSequences,
+            maxOutputTokens: this.maxOutputTokens,
+            temperature: this.temperature,
+            topP: this.topP,
+            topK: this.topK,
+          },
+        })
+      : new TextServiceClient({
+          authClient: new GoogleAuth().fromAPIKey(this.apiKey),
+        });
+  }
+
+  _llmType(): string {
+    return "googlegenerativeai";
+  }
+
+  /**
+   * Count tokens.
+   * @param input BaseLanguageModelInput - text or multimodal
+   * @returns Number of tokens
+   */
+
+  async countTokens(
+    input: BaseLanguageModelInput,
+    options?: this["ParsedCallOptions"]
+  ): Promise<number> {
+    if (!this._isGenerateContentModel) {
+      throw new Error(
+        `countTokens is not implemented for ${this.modelName} model`
+      );
+    }
+    const convertedInput = this._convertInputToGenerateContent(input);
+    const res = await this.caller.callWithOptions(
+      { signal: options?.signal },
+      async () => {
+        const output = await (this.client as GenerativeModel).countTokens(
+          convertedInput
+        );
+        return output;
+      }
+    );
+    return res.totalTokens;
+  }
+
+  _convertInputToGenerateContent(input: BaseLanguageModelInput): Part[] {
+    return convertInput(input, this.modelName.includes("vision"));
+  }
+
+  _generateContentResponseToText(response: EnhancedGenerateContentResponse) {
+    if (!response.candidates) {
+      return "";
+    }
+    const [message] = response.candidates;
+    const text = message.content.parts
+      .map((part) => part.text)
+      .filter(Boolean)
+      .flat(1)
+      .join(" ");
+    return text ?? "";
+  }
+
+  async invoke(
+    input: BaseLanguageModelInput,
+    options?: this["ParsedCallOptions"]
+  ) {
+    if (this._isGenerateContentModel) {
+      return super.invoke(input, options);
+    }
+    let prompt = this._convertInputToGenerateContent(input);
```

Comment:
Do we need to override `invoke` or can we just override `_call`? Or is this because we want to support message inputs?

It's a bit odd to specifically add support for messages into an LLM as a recommended flow (though it's of course possible) - I think my preference would be to keep LLMs text in/text out and then add chat model support for this.

Reply:
(I warned y'all about this months ago, I just want to remind you. 🙃)

Despite the misnomer of L"L"M, I'd want the non-chat version to be multimodal if possible.

I think when I was looking at this, however, I was looking at overriding `predictMessages()` instead, anticipating future multimodal responses as well.